pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexPlugin.java

```diff
@@ -0,0 +1,35 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+/**
+ * IndexPlugins are the way {@link IndexType}s are registered in a {@link IndexService}.
+ *
+ * In order to create an IndexService, a set of IndexPlugin must be provided. Although IndexTypes could be directly
+ * added into a IndexService, this indirection is used to decouple the way indexes are discovered
+ * (usually by {@link java.util.ServiceLoader} services) and the actual implementation.
+ *
+ * In order to mark a class as a {@link java.util.ServiceLoader} service, some metadata has to be added. Java modules
+ * define a typesafe way to define services, but given that Pinot does not use them right now, the easier way to create
```

Comment:
I am not very familiar with the ServiceLoader. but say in the future, we want to use the type. Will that be a problem? 
What's the implication of using Google AutoService? Compared with the other type safe method.

Reply:
Historically, in order to create a ServiceLoader service you need to add some text in the MANIFES file all jars usually have. This is tedious, error prone and personally I think that is one of the reasons most people don't know about ServiceLoader system. This example from [baeldung](https://www.baeldung.com/java-spi) shows how to manually create a ServiceLoader service. There is a new typesafe mechanism that was added when using Java modules. Given that we don't use it, I'm not going to explain that much about it, but it is important to say that if someday use Java modules, both ways to declare services are compatible.

[Baeldung](https://www.baeldung.com/google-autoservice) also has another tutorial showing how to use Google AutoServe.
 
What Google AutoService does is a very simple trick based on standards:
- You annotate a class with `@AutoService(WhateverInterfaceTheClassImplement.class)`
- AutoService uses a Java preprocessor, which is a standard `javac` feature that offers hooks where your code can apply. What the AutoService Java preprocessor does is trivial: It just generate the entry in the MANIFEST.MF file you would need to add manually instead.
- Given that this is only done in compile time, you don't even need to have the AutoService dependency at runtime.

So there is no magic here, no reflection, anything. It is just a typesafe way to generate an entry in the MANIFEST without having to add it manually.